doc: updated data type support tables for clarity (#3777)
uxlfoundation/oneDNN: doc/programming_model/data_types.md

The following uArchs and associated devices have specialized optimizations in the
library:
 * Xe-LP
   * Intel UHD Graphics for 11th-14th Gen Intel(R) Processors
   * Intel Iris Xe Graphics
   * Intel Iris Xe MAX Graphics (formerly DG1)
 * Xe-HPG
   * Intel Arc Graphics (formerly Achemist)
   * Intel Data Center GPU Flex Series (formerly Arctic Sound)
 * Xe-HPC
   * Intel Data Center GPU Max Series (formerly Ponte Vecchio)
 * Xe2-LPG
   * Intel Graphics for Intel Core Ultra processors (Series 2) (formerly Lunar Lake)
 * Xe2-HPG
   * Intel Arc B-Series Graphics (formerly Battlemage)
 * Xe3-LPG
   * Intel Arc graphics for future Intel Core Ultra processors (code name Panther Lake)

The following table indicates the data types support for each uArch supported by oneDNN.

| ISA      | f64    | f32    | bf16   | f16    | s8/u8  | f8     | f4     | s4/u4  |
| -------- | ------ | ------ | ------ | ------ | ------ | ------ | ------ | ------ |
| Xe-LP    |        | `+`    |        | `+`(1) |        |        |        |        |
| Xe-HPG   |        | `+`    | `+`    | `+`(1) |        |        |        |        |

> | Xe-HPC   | `+`    | `+`    | `+`    | `+`    | `+`    | `.`    |        |        |

Reviewer: It looks like support via upconversion here refers only to support through native upconversion, rather than more costly emulated upconversion. Is that accurate?

> | ISA      | f64     | f32     | bf16    | f16     | s8/u8   | f8      | f4_e2m1 | s4/u4   |
> | -------- | ------- | ------- | ------- | ------- | ------- | ------- | ------- | ------- |
> | Xe-LP    |         | `+`     |         | `+`(1)  |         |         |         |         |
> | Xe-HPG   |         | `+`     | `+`     | `+`(1)  |         |         |         | `.`     |
> | Xe-HPC   | `+`     | `+`     | `+`     | `+`     | `+`     | `.`     | `.`     | `.`     |
> | Xe2-LPG  | `+`     | `+`     | `+`     | `+`     | `+`     | `.`     | `.`     | `.`     |
> | Xe2-HPG  | `+`     | `+`     | `+`     | `+`     | `+`     | `.`     | `.`     | `.`     |
> | Xe3-LPG  | `+`     | `+`     | `+`     | `+`     | `+`     | `+`     | `.`     | `.`     |

Reviewer 2: Some corrections/update to this table:
```suggestion
| ISA      | f64     | f32     | bf16    | f16     | s8/u8   | f8      | f4_e2m1 | s4/u4   |
| -------- | ------- | ------- | ------- | ------- | ------- | ------- | ------- | ------- |
| Xe-LP    |         | `+`     | `.`     | `+`(1)  | `+`     |         |         |         |
| Xe-HPG   |         | `+`     | `+`     | `+`(1)  | `+`     | `.`     |         | `+`     |
| Xe-HPC   | `+`     | `+`     | `+`     | `+`     | `+`     | `.`     | `.`     | `+`     |
| Xe2-LPG  | `+`     | `+`     | `+`     | `+`     | `+`     | `.`     | `.`     | `+`     |
| Xe2-HPG  | `+`     | `+`     | `+`     | `+`     | `+`     | `.`     | `.`     | `+`     |
| Xe3-LPG  | `+`     | `+`     | `+`     | `+`     | `+`     | `+`     | `.`     | `+`     |
```

Reviewer 2: @vpirogov -- it depends on whether zp are present. With zp, int4 data will be upconverted to int8 prior to compute (to allow the zp to be applied pre-compute). If there are no zp, we often pass int4 data directly to dpas, since it supports arbitrary int{2,4,8} x int{2,4,8} combinations.

Author: Thanks! I'll leave it at `.` as we don't do any `int4` x `int4` math with systolic.

Legend:

> * `+` indicates full support with compute acceleration.
> * `.` indicates functional support through conversions to higher precision data
>   type or support as a storage format for weights.

Reviewer 2: I'm wondering if readers might come away from this to a misconception that `.` entries are not optimized. What about something like:
```suggestion
* `+` indicates oneDNN uses native HW compute support for this data type.
* `.` indicates oneDNN supports this data type via conversion to a higher precision data type.
```

Author: Thanks!

> Footnotes:
> 1. Xe-LP and Xe-HPG architectures do not support f16 operations with f32
>    accumulation. Consider using
>    [relaxed accumulation mode](@ref dev_guide_attributes_accumulation_mode)
>    for the best performance results.

Reviewer 2: This comment isn't correct as written. Suggest:
```suggestion
Footnotes:
1. Xe-LP architecture does not natively support f16 operations with f32
   accumulation. Consider using
   [relaxed accumulation mode](@ref dev_guide_attributes_accumulation_mode)
   for the best performance results.
```

Author: Looks good.

### Intel(R) Architecture Processors

oneDNN performance optimizations for Intel Architecture Processors are
specialized based on Instruction Set Architecture (ISA). The following
table indicates data types support for every supported ISA:

| ISA                                                  | f64    | f32    | bf16   | f16    | s8/u8  | f8     | f4     | s4/u4  |
| ---------------------------------------------------- | ------ | ------ | ------ | ------ | ------ | ------ | ------ | ------ |
| Intel SSE4.1                                         |        | `+`    |        |        |        |        |        |        |
| Intel AVX                                            |        | `+`    |        |        |        |        |        |        |
| Intel AVX2                                           |        | `+`    |        |        | `+`    |        |        |        |
| Intel AVX2 with Intel DL Boost (int8)                |        | `+`    |        |        | `+`(1) |        |        |        |
| Intel AVX2 with Intel DL Boost (int8, bf16)          |        | `+`    | `+`    |        | `+`    |        |        |        |

> | Intel AVX2 with Intel DL Boost (int8, bf16, fp16)    |        | `+`    | `+`    | `.`    | `+`    |        |        |        |

Author: Thanks, updated.

| Intel AVX-512                                        |        | `+`    | `.`(2) |        | `+`(1) |        |        |        |
| Intel AVX-512 with Intel DL Boost (int8)             |        | `+`    | `.`(2) |        | `+`    |        |        |        |
| Intel AVX-512 with Intel DL Boost (int8, bf16)       |        | `+`    | `+`    |        | `+`    |        |        |        |
| Intel AVX10.1/512                                    |        | `+`    | `+`    | `.`(3) | `+`    | `.`    |        |        |

> | Intel AVX10.1/512 with Intel AMX (int8, bf16)        |        | `+`    | `+`    | `.`(3) | `+`    | `.`    |        | `.`    |

Author: Sorted this one out. SPR ISA is functionally equivalent to Intel AVX10.1/512,  but does not have specific CPUID bit.
